File: scripts/create-test-items.ts
import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3';

type Args = {
  bucket?: string;
  prefix?: string;
  count: number;
  region?: string;
  endpoint?: string;
  concurrency: number;
  forcePathStyle: boolean;
};

function parseArgs(argv: string[]): Args {
  const args: Args = {
    count: 10005,
    concurrency: 25,
    forcePathStyle: false,
  };

  for (let i = 0; i < argv.length; i += 1) {
    const arg = argv[i];
    switch (arg) {
      case '--bucket':
        args.bucket = argv[i + 1];
        i += 1;
        break;
      case '--prefix':
        args.prefix = argv[i + 1];
        i += 1;
        break;
      case '--count':
        args.count = Number(argv[i + 1]);
        i += 1;
        break;
      case '--region':
        args.region = argv[i + 1];
        i += 1;
        break;
      case '--endpoint':
        args.endpoint = argv[i + 1];
        i += 1;
        break;
      case '--concurrency':
        args.concurrency = Number(argv[i + 1]);
        i += 1;
        break;
      case '--force-path-style':
        args.forcePathStyle = true;
        break;
      default:
        break;
    }
  }

  return args;
}

function normalizePrefix(prefix?: string): string {
  if (!prefix) {
    return '';
  }
  return prefix.endsWith('/') ? prefix : `${prefix}/`;
}

async function main() {
  const args = parseArgs(process.argv.slice(2));

  if (!args.bucket) {
    console.error('Missing required --bucket');
    process.exit(1);
  }

  if (!Number.isFinite(args.count) || args.count <= 0) {
    console.error('Invalid --count value');
    process.exit(1);
  }

  if (!Number.isFinite(args.concurrency) || args.concurrency <= 0) {
    console.error('Invalid --concurrency value');
    process.exit(1);
  }

  const prefix = normalizePrefix(args.prefix);
  const width = String(args.count).length;

  const client = new S3Client({
    region: args.region,
    endpoint: args.endpoint,
    forcePathStyle: args.forcePathStyle || undefined,
  });

  let completed = 0;
  const total = args.count;

  const uploadOne = async (index: number) => {
    const key = `${prefix}item-${String(index + 1).padStart(width, '0')}.txt`;
    await client.send(
      new PutObjectCommand({
        Bucket: args.bucket,
        Key: key,
        Body: '',
        ContentType: 'text/plain',
      })
    );
    completed += 1;
    if (completed % 500 === 0 || completed === total) {
      console.log(`Uploaded ${completed}/${total}`);
    }
  };

  for (let i = 0; i < total; i += args.concurrency) {
    const batch: Array<Promise<void>> = [];
    const end = Math.min(total, i + args.concurrency);
    for (let j = i; j < end; j += 1) {
      batch.push(uploadOne(j));
    }
    await Promise.all(batch);
  }
}

main().catch((error) => {
  console.error('Failed to create test items:', error);
  process.exit(1);
});

File: scripts/create-test-items.py
from __future__ import annotations

import argparse
import sys
import threading
from concurrent.futures import ThreadPoolExecutor, as_completed

import boto3
from botocore.config import Config


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Create empty test items in an S3 bucket.")
    parser.add_argument("--bucket", required=True)
    parser.add_argument("--prefix", default="")
    parser.add_argument("--count", type=int, default=10005)
    parser.add_argument("--region", default=None)
    parser.add_argument("--endpoint", default=None)
    parser.add_argument("--concurrency", type=int, default=25)
    parser.add_argument("--force-path-style", action="store_true", dest="force_path_style")
    return parser.parse_args()


def normalize_prefix(prefix: str) -> str:
    if not prefix:
        return ""
    return prefix if prefix.endswith("/") else f"{prefix}/"


def main() -> int:
    args = parse_args()

    if args.count <= 0:
        print("Invalid --count value", file=sys.stderr)
        return 1
    if args.concurrency <= 0:
        print("Invalid --concurrency value", file=sys.stderr)
        return 1

    prefix = normalize_prefix(args.prefix)
    width = len(str(args.count))
    total = args.count

    s3_config = Config(
        s3={"addressing_style": "path"} if args.force_path_style else {},
        max_pool_connections=max(args.concurrency, 10),
    )
    client = boto3.client(
        "s3",
        region_name=args.region,
        endpoint_url=args.endpoint,
        config=s3_config,
    )

    completed = 0
    lock = threading.Lock()

    def upload_one(index: int) -> None:
        nonlocal completed
        key = f"{prefix}item-{str(index + 1).zfill(width)}.txt"
        client.put_object(Bucket=args.bucket, Key=key, Body=b"", ContentType="text/plain")
        with lock:
            completed += 1
            if completed % 500 == 0 or completed == total:
                print(f"Uploaded {completed}/{total}")

    with ThreadPoolExecutor(max_workers=args.concurrency) as executor:
        futures = [executor.submit(upload_one, i) for i in range(total)]
        for future in as_completed(futures):
            future.result()

    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as error:
        print(f"Failed to create test items: {error}", file=sys.stderr)
        sys.exit(1)
